main/java/com/bchupika/user/repository/UserRepository.java

```diff
@@ -0,0 +1,13 @@
+package com.bchupika.user.repository;
+
+import com.bchupika.user.dto.UserRegistrationDto;
+import com.bchupika.user.entity.User;
+import com.bchupika.user.exception.DataProcessingException;
+
+import java.util.Optional;
+
+public interface UserRepository {
+    Number save(UserRegistrationDto user) throws DataProcessingException;
```

Comment:
Why return type is `Number`? Why not `User`?

Also it is better to work with entities on the repository level, at least accept entity `User` instead of `UserRegistrationDto` as an method argument.

Also,  DataProcessingException is unchecked exception because you declared it in such a way:`public class DataProcessingException extends RuntimeException`.

Unchecked exceptions should not be declared using `throws` keyword. So expected method signature could be:
```suggestion
    User save(User user);
```